Add seriestracker

MaximeWack opened the pull request:
Upstream is github:maximewack/seriestracker

### Brief summary of what the package does

seriestracker implements a major mode for tracking TV shows.
TV shows data (episode list, release dates, etc.)
are sourced from the free database hosted at episodate.com
The mode presents an outlined list of tracked shows,
their episodes and release dates, and enables the user
to see when new episodes for their favorite shows get released,
and track their progress in watching a series.

### Direct link to the package repository

https://github.com/MaximeWack/seriestracker

### Your association with the package

I am the author and maintainer of this package

### Relevant communications with the upstream package maintainer

None needed

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

purcell commented:
Thanks for this, it looks like it could be useful. Some notes after a quick look through:

- Your package link above is wrong, but I found the repo anyway.
- The camel case name of the package is highly unconventional - use `series-tracker` (or perhaps `seriestracker`) instead.
- The `st--` and `st-` prefixes are inappropriate: use `series-tracker-` and `series-tracker--` instead, to match the package name. `package-lint` should have told you about this when you ran it?
- Stylistically, all those blank lines are also very unconventional and make the code hard to read for experienced emacs lispers.
- Regarding [this code](https://github.com/MaximeWack/seriesTracker/blob/0d3e60c9d20f6102bd1086fcd2f987b5ae8736f4/seriesTracker.el#L870-L881), don't use ad-hoc face specifiers: declare proper faces instead, and `:inherit` from builtins like `success`/`error`/`warning` etc. by default so that the results will automatically look reasonable in most themes. Making local colour choices is a recipe for readability issues in anything but your own config.
- Don't use lots of `local-set-key`, e.g. [here](https://github.com/MaximeWack/seriesTracker/blob/0d3e60c9d20f6102bd1086fcd2f987b5ae8736f4/seriesTracker.el#L1125-L1146) - instead, define a keymap and associate it with your mode. The `define-derived-mode` form will create a keymap var for you, which you can modify in the following code.

Hope that helps!

MaximeWack commented:
Hi,

Thank you very much for the feedback.

- sorry for the link, I had to fill the form twice and I must have overlooked the url the second time. I edited the form to correct it
- I renamed the whole package to `st` for consistency and brevity
- `package-lint` did warn me about the naming convention. I must confess I willingly ignored it as the function naming convention still was consistent, beside the name of the package, that I was reluctant to change because of the name of the repo and already publishing it. I finally have changed the name to `st` (and github does the right thing of linking the old url to the new so everything is ok). I have fixed further warnings from `package-lint`.
- I have removed the superfluous blank lines
- dealing in a cleaner way with faces was already on my todo list, but I guess its priority just went up ^^
- dealing with the keymap also was on my list and gets a bump.

Is it okay if I update this PR with a recipe that points to the `develop` branch so that I don't have to release to `master` for every update? When everything is correct I will push one last commit reverting to the default branch.

This is my first emacs package and I'm still learning so I apologize for all the unconventional oddities.
Thank you for your kind feedback! :-)

MaximeWack commented:
The latest commits in `develop` address the last two points regarding the faces and the keymap definition.

I hope my corrections are satisfactory! Don't hesitate to point out any irregularity, I'm willing to learn the colloquial way of doing things.^

riscy commented:
I'll hop in with a couple answers and some feedback, too.

The package, variable names, and function names should all use the same prefix (in this case your package is still unconventionally named, and your new function/variable prefixes don't match).  MELPA doesn't care about the name of the repository, but sometimes it's good if it matches.  As for the name itself, arguably `st` is a little on the short and ambiguous side.  It may feel inconvenient, but this is a simple and fair way to prevent namespace collisions across MELPA (and hopefully elsewhere).

> Is it okay if I update this PR with a recipe that points to the develop branch so that I don't have to release to master for every update?

If it's helpful for you, I can check your `develop` branch instead of your default branch while this PR is open.  That'll be easier than you changing this PR's recipe file.  In general we prefer it if MELPA can index the latest version of your code on the current default branch, and we use MELPA stable (q.v.) if you want to be explicit about tagging release versions.

That said, here's some quick automated feedback on the current `develop` branch:

byte-compile (using Emacs 27.1):
```
seriesTracker.el:188:31:Warning: assignment to free variable `series-index'
seriesTracker.el:195:15:Warning: reference to free variable `series-index'
In st--refresh:
seriesTracker.el:327:14:Warning: reference to free variable `st--fold-cycle'
In st-next-line:
seriesTracker.el:440:4:Warning: `next-line' is for interactive use only; use
    `forward-line' instead.
In end of data:
seriesTracker.el:1153:1:Warning: the following functions are not known to be
    defined: acons, first-next-date
```
- 651:0: error: `st-transient-variable:choice' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions). (from package-lint; this one may have been lost in the prefix noise)
- seriesTracker.el#L608: Could this point/point-max comparison use `eobp`?
- seriesTracker.el#L286: Could this point/point-min comparison use `bobp`?

MaximeWack commented:
Hi Chris,

Thanks for the feedback!

I renamed the whole project, functions, and vars to `seriestracker`, and updated the PR text as well as the links and repo name.
`master` is the stable branch and `develop` the current default branch so it's best if this recipe points to `develop`. I have added a commit reflecting this and the name change to this PR.  
Tell me if you'd prefer I squash these two commits together.

I fixed all remaining byte-compile and package-lint warnings (except for the one with `next-line`, as using `forward-line` introduces undesired behavior while `next-line` does exactly what is meant to be done by the command, so I'm not sure I should use `forward-line` despite the warning?)
  
I followed your advice and used `eobp` and `bobp`, thanks!

For the non-standard colon separator I only did as `magit` does ^. -  (and fixed it)

I hope the naming convention is now acceptable (I'm not sure I understood what you referred to by "your new function/variable prefixes don't match" as everything was uniformly prefixed with "st-"?)

Lastly, I'm still having difficulties with… autoloading I think?

With this package installed, upon startup all functions named `seriestracker-` are available, all `seriestracker--` ones are not loaded, but `seriestracker` is also inexistant before any of the `seriestracker-` is called.  
Per the manual I was expecting the exact opposite by using the `;;;###autoload` magic comment before `(defun seriestracker)` (ie only `seriestracker` would be visible and calling it would load the rest of the package).  
Removing the magic comment doesn't change anything and the `seriestracker` function refuses to load unless I call one of the other functions, `(require 'seriestracker)`, or `:demand` the package in `use-package`.

What would I need to do to obtain the desired behavior? (or at least have the `seriestracker` function available on startup?)
(I hope this is not totally out of the scope of this PR!)

Thank you for your help!

purcell commented:
Quick drive-by comments here:

MELPA would much rather track your default branch, ie. whatever github shows by default. This prevents an issue where developers change their "git flow" (or similar) workflow later and start working on "master" or "main". So I'd suggest just developing on `master`.

Re. the autoload, you've defined that exactly right for the behaviour you expected. Note that `package-initialize` is usually what arranges for the the installed packages' autoloads to be evaluated. If you've installed the correct version of the package you can check its `seriestracker-autoloads.el` file.

riscy commented:
Just adding to the above --

> using forward-line introduces undesired behavior while next-line does exactly what is meant to be done by the command, so I'm not sure I should use forward-line despite the warning?

I think in your case `line-move` may be a better candidate.  The behavior of `next-line` can be customized by the user in a way that may not be what your program expects.

A very minor checkdoc pops up:
- seriestracker.el:919: First sentence should end with punctuation

So I think the current remaining issues are to (a) set the recipe to track the default branch and (b) have another run over the checklist to make sure the above items are caught and handled.

MaximeWack commented:
I've switched back to developing on master and tagging releases, and updated the PR as such.

`line-move` works as intended! Thanks!  
I also corrected that last docstring warning.

Everything should be ok now, I hope I have not overlooked anything.

purcell commented:
Looks great, merging now! My only small observation is the naming of the `seriestracker--getJSON` function - you should prefer snake-case there.

Thanks for your patience, and welcome to MELPA! :tada:

MaximeWack commented:
Great!
Thanks for _your_ patience, kindness, and helping me promote my package to the quality standards of MELPA.
Your work in maintaining this archive is very much appreciated and I'm happy to contribute!
